### PR TITLE
[Feature] `SamplerWithoutReplacement` state dictionary

### PR DESCRIPTION
**Pull request by matteobettini**
TODO:
- [x] Tests

### CODE REVIEW

**Review by vmoens — APPROVED** [review]
LGTM thanks so much for this

**Comment by vmoens** [comment]
There is some torchsnapshot import that needs to be fixed

**Comment by matteobettini** [comment]
Could you link me the problem in the CI?

I am not able to find it.
(I only see another problem with a vec env which is unrelated to this)

File: .github/unittest/linux_olddeps/scripts_gym_0_13/environment.yml [hunk 1]
```diff
@@ -25,3 +25,4 @@ dependencies:
     - patchelf
     - pyopengl==3.1.4
     - ray<2.8.0
+    - torchsnapshot
```
Comment (vmoens) [inline .github/unittest/linux_olddeps/scripts_gym_0_13/environment.yml:28]: Nope let's avoid installing it everywhere, let's just skip tests where it is not installed already

**Review by vmoens — COMMENTED** [review]
We already have pipelines where snapshot is installed and tested. Let's avoid installing it anywhere else!

File: test/test_rb.py [hunk 2]
```diff
@@ -1553,6 +1559,53 @@ def test_error_noninit(self):
 
 
 class TestSamplers:
+    @pytest.mark.parametrize("backend", ["torch", "torchsnapshot"])
+    def test_sampler_without_rep_state_dict(self, backend):
+        if backend == "torchsnapshot" and not _has_snapshot:
+            pytest.skip()
```
Comment (vmoens) [inline test/test_rb.py:1565]: Otherwise you could make the backend list dependent of whether _has_torchsnapshot is true of false

Comment (matteobettini) [reply test/test_rb.py]: if you prefer that...

In that case it doesn't show that there were tests skipped and the number of tests will vary depending on the pipeline

Comment (vmoens) [reply test/test_rb.py]: Yes but that's better, for instance fbcode does not like skipping tests 
Better not have a test listed than skipped systematically

**Review by vmoens — APPROVED** [review]
LGTM! Thanks
Can you edit the description?

**Pull request merged**